File: frontend/src/services/economico/Ingresos.ts
import API_BASE from "../../config/api";

export type IngresoEconomico = {
  id: number;
  fecha: string;
  fuente: string;
  tipo: "Venta producción" | "Subvención" | "Préstamo" | "Otro";
  monto: number;
  responsable: string;
  descripcion: string;
};

export type NuevoIngresoEconomico = Omit<IngresoEconomico, "id">;

const EMPRESA_ID = 1;
const GRANJA_ID = 1;

function mapApiToIngreso(api: any): IngresoEconomico {
  return {
    id: api.id,
    fecha: api.fecha,
    fuente: api.fuente,
    tipo: api.tipo,
    monto: api.monto,
    responsable: api.responsable,
    descripcion: api.descripcion ?? "",
  };
}

export async function getIngresos(): Promise<IngresoEconomico[]> {
  const res = await fetch(
    `${API_BASE}/economico/ingresos/?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`
  );
  if (!res.ok) {
    throw new Error("Error al obtener ingresos económicos");
  }
  const data = await res.json();
  return data.map(mapApiToIngreso);
}

export async function addIngreso(
  ing: NuevoIngresoEconomico
): Promise<IngresoEconomico> {
  const payload = {
    empresa_id: EMPRESA_ID,
    granja_id: GRANJA_ID,
    ...ing,
  };
  const res = await fetch(`${API_BASE}/economico/ingresos/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) {
    throw new Error("Error al crear ingreso económico");
  }
  const data = await res.json();
  return mapApiToIngreso(data);
}

export async function updateIngreso(
  id: number,
  ing: NuevoIngresoEconomico
): Promise<IngresoEconomico> {
  const res = await fetch(
    `${API_BASE}/economico/ingresos/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(ing),
    }
  );
  if (!res.ok) {
    throw new Error("Error al actualizar ingreso económico");
  }
  const data = await res.json();
  return mapApiToIngreso(data);
}

export async function deleteIngreso(id: number): Promise<void> {
  const res = await fetch(
    `${API_BASE}/economico/ingresos/${id}?empresa_id=${EMPRESA_ID}&granja_id=${GRANJA_ID}`,
    { method: "DELETE" }
  );
  if (!res.ok) {
    throw new Error("Error al eliminar ingreso económico");
  }
}

File: backend/crud/economico/Ingresos.py
// src/services/economicoIngresosService.ts
export type IngresoEconomico = {
  id: string;
  fecha: string;
  fuente: string;
  tipo: "Venta producción" | "Subvención" | "Préstamo" | "Otro";
  monto: number;
  responsable: string;
  descripcion: string;
};

const STORAGE_KEY = "economico_ingresos";

function getAll(): IngresoEconomico[] {
  const raw = localStorage.getItem(STORAGE_KEY);
  return raw ? JSON.parse(raw) : [];
}

function saveAll(lista: IngresoEconomico[]) {
  localStorage.setItem(STORAGE_KEY, JSON.stringify(lista));
}

export function getIngresos() {
  return getAll();
}

export function addIngreso(ing: Omit<IngresoEconomico, "id">) {
  const nuevo: IngresoEconomico = { ...ing, id: crypto.randomUUID() };
  const lista = getAll();
  lista.push(nuevo);
  saveAll(lista);
  return nuevo;
}

export function updateIngreso(id: string, ing: Omit<IngresoEconomico, "id">) {
  const lista = getAll();
  const idx = lista.findIndex(i => i.id === id);
  if (idx >= 0) {
    lista[idx] = { ...ing, id };
    saveAll(lista);
    return lista[idx];
  }
  throw new Error("Registro no encontrado");
}

export function deleteIngreso(id: string) {
  const lista = getAll().filter(i => i.id !== id);
  saveAll(lista);
}
